### ISSUE TITLE
improve experience when configuring an alternative `JobRepository` 

### ISSUE THREAD

**Issue by joshlong** [spring-projects/spring-batch#4718]
improve experience when configuring an alternative `JobRepository`, as the DefaultBatchConfiguration, upon which Spring Boot in turn relies, assumes only JDBC-based infrastructure. This means that, when using Spring Boot, you can't simply replace one bean of type `JobRepository` and have things work. Everything must be redone from scratch, as we used to do before Spring Boot existed.  

This will be a problem for users of Spring Batch who want to avail themselves of the new "resourceless" and MongoDB job repository alternatives.

**Comment by fmbenhassine** [comment]
`@EnableBatchProcessing` (and in turn `DefaultBatchConfiguration`) was designed as the Java counterpart of the XML element `job-repository` from the batch namespace where a datasource is required:

```xml
<batch:job-repository id="jobRepository" data-source="dataSource" transaction-manager="transactionManager"/>
```

So yes, `@EnableBatchProcessing` assumes a JDBC infrastructure, and this should change now that we introduced the mongo job repository and the resourceless job repository.

I have a couple ideas of how we can address this issue. I will share them with the community when we start working on this for the next major. This comment is only to validate and acknowledge the issue. Thank you for reporting it, @joshlong !

**Comment by fmbenhassine** [comment]
@krewetka which shutdown problem are you referring to? How is that related to this issue? Please open a separate [discussion](https://github.com/spring-projects/spring-batch/discussions) to explain your problem and I will be happy to help there.

FTR, v6 is planned for November this year and there is no plan for v5.3 at the moment. In the meantime and for those who want to use a non-jdbc job repository with v5.2, please check this [SO thread](https://stackoverflow.com/a/79492398/5019386) in which I attached a complete example.

**Comment by krewetka** [comment]
Hi, sorry my bad. I am referring to https://github.com/spring-projects/spring-batch/issues/4728#issuecomment-2578356238 which was marked as duplicate of this issue.  Should I still open another one?🤔

We have mongo setup running properly (but with H2 still there).

We did try to exclude DataSourceAutoConfiguration.class but then we were still facing some issues but let me try again one more time and see the results.

**Comment by fmbenhassine** [comment]
@krewetka 

> Should I still open another one then?

Yes please. The shutdown lifecycle issue is different than requiring h2 for a mongo setup.

**Comment by fmbenhassine** [comment]
@krewetka Thank you for your feedback, really appreciated! I understand, that's confusing. And I must admit that the default configuration of Spring Batch is not ideal, especially after the introduction of the resourceless job repository and the mongo job repository. I believe the introduction of these two new implementations has made the issues with previous design decisions more visible (ie the fact the default batch configuration assumes a JDBC infrastructure, the requirement of a `JobExplorer`, etc).

I started working on the next major release and wanted to share a few things:

- When using a `ResourcelessJobRepository`, there are no metadata to explore (there is no `ResourcelessJobExplorer`, this does not make sense). Therefore, the default batch configuration should not require a `JobExplorer`. This was fixed in https://github.com/spring-projects/spring-batch/issues/4825
- With Spring Boot, using `ResourcelessJobRepository` or `MongoJobRepository` should not require the user to exclude `DataSourceAutoConfiguration` or have to add a dependency to H2 or HSQLDB. Therefore, I think the resourceless job repository is a better default than the JDBC one in Spring Boot. I will discuss that with the Boot team.
- `EnableBatchProcessing` should not be tied to a JDBC infrastructure (this is historical, see my [previous comment](https://github.com/spring-projects/spring-batch/issues/4718#issuecomment-2514623019)). Therefore, all attributes related to the configuration of the JDBC job repository should be moved elsewhere. My idea is to follow the same approach of configuring Spring Security's session store with `@EnableJdbcHttpSession`, `@EnableRedisHttpSession`, etc. We could have the same for Spring Batch with `@EnableJdbcJobRepository`, `@EnableMongoJobRepository`, etc where each annotation have specific attributes to configure the metadata store.

And more generally, and after gathering feedback from the community over the years, I believe we need to simplify the API "surface" that batch users have to deal with. Currently, there are many different APIs to do the same things, which is confusing. I will open a separate issue for that with more details.
